src/main/scala/edu/colorado/plv/cuanto/scoot/concrete_interpreter/Interpreter.scala

```diff
@@ -0,0 +1,84 @@
+package edu.colorado.plv.cuanto.scoot
+package concrete_interpreter
+
+import edu.colorado.plv.cuanto.scoot.jimple._
+import soot.jimple.internal.{JAssignStmt, JReturnStmt, JimpleLocal}
+//import edu.colorado.plv.cuanto.scoot.concrete_interpreter.ConcreteMemory.{CValue,CInteger}
+
+import scala.collection.immutable.HashMap
+import scala.util.{Success, Try}
+
+/** Implement an interpreter for sequences of Jimple assignment
+  * statements that represent integer arithmetic programs
+  */
+object Interpreter {
+
+  /** An "execution environment" or state, mapping variables (of type
+    * `Local`) to integer values */
+  type Env = Map[String,Int]
+
+  /** An environment with no assigned variables */
+  val emptyEnv: Env = new HashMap[String,Int]()
+
+  /** Interpret arithmetic expressions encoded as a single `Value` */
+  def denote(v: Value, env: Env = emptyEnv): Option[Int] = v match {
+    case Local(s) => env get s
+    case IntConstant(v) => Some(v)
+    case AddExpr(e1, e2) => for {
+      arg1 <- denote(e1, env)
+      arg2 <- denote(e2, env)
+    } yield arg1 + arg2
+    case SubExpr(e1, e2) => for {
+      arg1 <- denote(e1, env)
+      arg2 <- denote(e2, env)
+    } yield arg1 - arg2
+    case MulExpr(e1, e2) => for {
+      arg1 <- denote(e1, env)
+      arg2 <- denote(e2, env)
+    } yield arg1 * arg2
+    case DivExpr(e1, e2) => for {
+      arg1 <- denote(e1, env)
+      arg2 <- denote(e2, env)
+    } yield arg1 / arg2
+    case NegExpr(e) => for {
+      arg <- denote(e, env)
+    } yield -arg
+  }
+  def interpretBody(b : Body): Try[CValue] = {
+    Try(internal_interpretBody(List(emptyEnv), b.getFirstNonIdentityStmt, b).getOrElse(throw new RuntimeException("interpreter exception")))
+  }
+  def updateEnv(env: List[Env], varname: Value, value: Int): List[Env] = {
+
+    val newEnv: List[Env] = varname match{
+      case Local(n) =>  {
+        List(env.head + ((n, value))) ::: env.tail
+      }
+      case _ => {
+        ???
+      }
+    }
+    newEnv
+  }
+  def internal_interpretBody(env : List[Env], loc: Stmt, b : Body): Option[CValue] = {
```

Comment:
Should this be `private` or `protected` if it's internal?

Reply:
I made everything private that shouldn't be used by outside users of the interpreter except the method "evaluate_expr(v: Value, env: Env = emptyEnv): Option[Int]" since it needs to be tested.  We should look into a lower level of accessibility that lets it be unit tested but not externally utilized.